= Pipeline Workflow
:fasta-index: https://github.com/guigolab/grape-nf/blob/master/templates/fastaIndex/SAMtools
:gem-index: https://github.com/guigolab/grape-nf/blob/master/templates/index/GEM
:star-index: https://github.com/guigolab/grape-nf/blob/master/templates/index/STAR
:rsem-index: https://github.com/guigolab/grape-nf/blob/master/templates/t_index/RSEM

- validate parameters and print summary
- check input datasets for runs and samples and detect if merge needed
- preprocess list of input files to detect paired-end/single-end and FASTQ/BAM
- prepare reference genome and annotation
- spawn processes depending on pipeline options (see details below)
- collect ouput files and write result list on pipeline db


== Process details

[cols="10l,90"]
|===
| fastaIndex     | create the fasta index file containing the chromosome lengths
| index          | create the reference index file(s) required by the selected mapper
| t_index        | create a transcriptome index required by some quantification programs
| mapping        | aligns the reads using the selected mapper
| mergeBam       | merge mapping BAM files if needed
| inferExp       | infer the experiment type from the mappping BAM files
| bigwig         | produce bigwig files
| contig         | produce BED files containing the contigs
| quantification | quantify transcripts and genes using the selected quantification program
|===

=== Fasta index

The `fastaIndex` process gets input from two channels. Both channels emit sets with two fields, a species identifier and a `file` object.
The first channel contains information about the reference genome while the second channel contains information from the reference
annotation.

The same format is used for the output channel, where the input species identifier and the `file` object for the fasta index are stored
as a set.

The process template can be viewed {fasta-index}[here^].

== Genome index

The `index` process uses the same input as the `fastaIndex` process.

The same format is used for the output channel, where the input species identifier and the `file` object for the mapping index folder
are stored as a set.

The process template is specific to the selected mapper and can be one of the following:

- {gem-index}[GEM^]
- {star-index}[STAR^]

=== Transcriptome index

The `t_index` process uses the same input as the `fastaIndex` process.

The same format is used for the output channel, where the input species identifier and the `file` object for the transcriptome index
folder are stored as a set.

The process is only run if the selected quantification program requires a transcriptome index. For now only RSEM requires it and here
it is the template:

- {rsem-index}[RSEM^]

=== Mapping
=== Merge BAM files
=== Infer experiment
=== BigWig
=== Contig
=== Quantification
